include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2012-2017, Association Scientifique pour la Geologie et ses
+ * Applications (ASGA). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of ASGA nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ASGA BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+
+//#include <memory>
+
+//#include <geogram/mesh/mesh.h>
+//#include <geogram/mesh/mesh_io.h>
+
+//#include <ringmesh/geogram_extension/geogram_extension.h>
+
+#include <ringmesh/basic/frame.h>
+//#include <ringmesh/mesh/mesh.h>
+//#include <ringmesh/mesh/mesh_index.h>
+
+namespace RINGMesh
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( CartesianGridBuilder );
+}
+
+/*!
+ * @file Generic class for creating cartesian grids
+ * @author Melchior Schuh-Senlis
+ */
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API CartesianGrid
+    {
+        friend class CartesianGridBuilder< DIMENSION >;
+
+    public:
+        CartesianGrid( intvecn< DIMENSION > nb_cells,
+            index_t nb_total_cells,
+            ReferenceFrame< DIMENSION > vec_cartesian_axis )
+            : nb_cells_( nb_cells ), cartesian_frame_( vec_cartesian_axis )
```

Comment:
This is a more advanced comment. You are doing a copy from vec_cartesian_axis  to cartesian_frame_, you could use the move operator from c++11 to avoid this: cartesian_frame_( std::move( vec_cartesian_axis ) ). This way, the variable vec_cartesian_axis  does not exist anymore and its memory is used for the cartesian_frame_ variable.